printing/proof_diffs.ml

```diff
@@ -317,31 +333,244 @@ let hyp_list_to_pp hyps =
+    | GEvar (n, l), nt' ->
+      (* pass down the old goal id *)
+      match_goals_r nt' nt' (Id.to_string n)
+    | GPatVar kind, GPatVar kind2 -> ()
+    | GApp (f,args), GApp (f2,args2) ->
+      glob_constr_g chg f f2;
+      List.iter2 (glob_constr_g chg) args args2
+    | GLetIn (na,b,t,c), GLetIn (na2,b2,t2,c2) ->
+      glob_constr_g chg b b2;
+      opt_glob_constr_g chg t t2;
+      glob_constr_g chg c c2
+    | GProd (na,bk,t,c), GProd (na2,bk2,t2,c2) ->
+      glob_constr_g chg t t2;
+      glob_constr_g chg c c2
+    | GLambda (na,bk,t,c), GLambda (na2,bk2,t2,c2) ->
+      glob_constr_g chg t t2;
+      glob_constr_g chg c c2
+    | GCases (sty,rtntypopt,tml,eqns), GCases (sty2,rtntypopt2,tml2,eqns2) ->
+      opt_glob_constr_g chg rtntypopt rtntypopt2;
+      List.iter2 (fun x x2 -> let (a,_) = x and (a2,_) = x2 in
+                 glob_constr_g chg a a2)
+          tml tml2;
+      List.iter2 (fun x x2 -> let (idl, patl, xa) = x.v and (id2, pat2, xa2) = x2.v in
+                 glob_constr_g chg xa xa2) eqns eqns2
+    | GLetTuple (nal,(na,typopt),tm,b), GLetTuple (nal2,(na2,typopt2),tm2,b2) ->
+      opt_glob_constr_g chg typopt typopt2;
+      glob_constr_g chg tm tm2;
+      glob_constr_g chg b b2
+    | GIf (c,(na,typopt),b1,b2), GIf (c2,(na2,typopt2),b12,b22) ->
+      glob_constr_g chg c c2;
+      opt_glob_constr_g chg typopt typopt2;
+      glob_constr_g chg b1 b12;
+      glob_constr_g chg b2 b22
+    | GRec (fk,idv,blv,tyv,bv), GRec (fk2,idv2,blv2,tyv2,bv2) ->
+      Array.iter2 (fun x x2 ->
+            List.iter2 (fun y y2 -> let (_,_,a,b) = y and (_,_,a2,b2) = y2 in
+                           opt_glob_constr_g chg a a2; glob_constr_g chg b b2)
+            x x2)
+          blv blv2;
+      Array.iter2 (glob_constr_g chg) tyv tyv2;
+      Array.iter2 (glob_constr_g chg) bv bv2
+    | GSort s, GSort s2 -> ()
+    | GHole (e,naming,_), GHole (e2,naming2,_) -> ()
+    | GCast (c, c'), GCast (c2, c'2) ->
+      glob_constr_g chg c c2;
+      (match c', c'2 with
+      | CastConv a, CastConv a2
+      | CastVM a, CastVM a2
+      | CastNative a, CastNative a2
+        -> glob_constr_g chg a a2
+      | CastCoerce, CastCoerce
+        -> ()
+      | _, _ -> raise (Diff_Failure "Unable to match goals betwen old and new proof states (2)"))
+    | GProj (p, c), GProj (p2, c2) ->
+      glob_constr_g chg c c2
+    | _, _ -> raise (Diff_Failure "Unable to match goals betwen old and new proof states (3)")
+    end;
+    if print then
+      indent := !indent - 2
+  in
+
+  (match ot with
+  | Some ot -> match_goals_r ot nt ""
+  | None -> ());
+  !nid_to_oid
+
+
+(* based on the code in show_proof *)
+let to_constr p =
+  let open CAst in
+  let pprf = Proof.partial_proof p in
+  (* todo: When does pprf have multiple elements?  Can it have zero elements? *)
+  let t = List.hd pprf in
+  let sigma, env = Pfedit.get_current_context ~p () in
```

Comment:
I'm pretty sure zero elements is not possible here.  If it's possible to have multiple elements, I'd like to know how to generate such a case so I can try it.

Reply:
@ejgallego can probably answer this. I think there is support for multiple proofs that is only used by the Derive plugin AFAIR.

Reply:
Thanks.  BTW, I tried `Set Nested Proofs Allowed.' and entering a new Goal but I still didn't get multiple entries in pprf.

Reply:
It might be for
~~~coq
Lemma foo : A with bar : B.
Proof. exact A_val. exact B_val. Qed.
~~~
I didn't check so this is just a conjecture.

Reply:
Ah, can you send me the exact text to try?

![image](https://user-images.githubusercontent.com/1253341/44080165-12636428-9f60-11e8-8be6-a23c87a2f6c6.png)

Reply:
``` coq
Variables A B : Set.
Inductive tree : Set := node : A -> forest -> tree

with forest : Set :=
| leaf : B -> forest
| cons : tree -> forest -> forest.

Theorem tree_size : tree -> nat with forest_size : forest -> nat.
```

Reply:
Thanks for the example text.  I still find pprf has only one element.  Perhaps it is a vestige and the code no longer generates pprf with more than one element.

Reply:
It seems that the ability to have several goals in `partial_proof` is not used in general. It is used in the `derive` plugin but I did not investigate more.